### ISSUE TITLE
Spike: Protobuf vs Flatbuffer Performance Evaluation

### ISSUE THREAD

**Issue by tustvold** [influxdata/influxdb_iox#2738]
Prior to #2724 we want to quantify the performance differential between the two to quantify what we might lose by making the switch.

In particular the following should be considered and where possible measured:

* Flatbuffer verification speed
* Protobuf decode speed
* Compaction throughput
* Write buffer throughput

Previously the performance of JSON vs flatbuffers was evaluated - https://github.com/influxdata/influxdb_iox/pull/606

**Comment by alamb** [comment]
I personally suggest measuring Compaction throughput and write buffer throughput (as a proxy for recovery speed) is all that matters -- the flatbuffer verification / protobuf decode speed is a secondary concern (as it is included in the other two measurements)

**Comment by tustvold** [comment]
> the flatbuffer verification / protobuf decode speed is a secondary concern (as it is included in the other two measurements)

Indeed, I only split them up to allow confirming if there is an order of magnitude difference before spending the time to plumb the protobuf into the write path. Currently the write protocol is converted to a flatbuffer entry and so it wouldn't be an entirely fair comparison

**Comment by alamb** [comment]
When you say "Protobuf Binary Write Protocol" is that referring to https://github.com/influxdata/influxdb-pb-data-protocol that @jacobmarble  worked on ? 

If so, I think the idea if 'switching to use protobuf after adding a dictionary encoding mechanism' makes all sorts of sense to me. It continues the high level idea from @pauldix  (and I think Jacob) to push some of the required ingest computation to the edge. 👍

**Comment by tustvold** [comment]
> When you say "Protobuf Binary Write Protocol" is that referring to https://github.com/influxdata/influxdb-pb-data-protocol

Yes, although it is perhaps worth noting these benchmarks are not doing any of the optimisations listed there as I wasn't aware of them - I just read the proto spec :laughing: The performance difference of these is likely to be incredibly marginal if anything, so I'm confident these benchmarks still stand

**Comment by tustvold** [comment]
@jacobmarble Correct
